### PR TITLE
optimize names for new waypoints (related to #9041)

### PR DESCRIPTION
**Pull request by moving-bits**
Optimizes name handling for new waypoints:
- uses a short name for new waypoints (as described in https://github.com/cgeo/cgeo/issues/9041#issuecomment-697926825, "new WP default name")
- counter for new waypoints is no longer wp type-specific, but "global" (within this cache)
- detecting the highest given number is no longer limited to 30 but done with a regular expression, so if a wp is named "Stage 47", the next waypoint will get index 48.
- for comparison old + new name scheme are used to stay compatible

I did not swap the order of name and number, as this would either create confusion or require a migration of the existing waypoints. Let's see, if the shortened default names are sufficient for visibility in window titles already.

### CODE REVIEW

**Comment by moving-bits** [comment]
@fm-sys @eddiemuc (don't remember who had been working on the PCN load/store functions)

I added comparison with the new `WaypointType.getNameForNewWaypoint()` to both `Waypoint.parseWaypointType()` and `Geocache.findWaypoint()` - pleas cross-check if that will work as expected. Thx.

**Comment by eddiemuc** [comment]
> @fm-sys @eddiemuc (don't remember who had been working on the PCN load/store functions)
> 
> I added comparison with the new `WaypointType.getNameForNewWaypoint()` to both `Waypoint.parseWaypointType()` and `Geocache.findWaypoint()` - pleas cross-check if that will work as expected. Thx.

I left a comment on this

**Comment by moving-bits** [comment]
@eddiemuc 

> > @fm-sys @eddiemuc (don't remember who had been working on the PCN load/store functions)
> > I added comparison with the new `WaypointType.getNameForNewWaypoint()` to both `Waypoint.parseWaypointType()` and `Geocache.findWaypoint()` - pleas cross-check if that will work as expected. Thx.
> 
> I left a comment on this

Thanks, but where can I find that comment?

**Comment by eddiemuc** [comment]
Seems I forgot to submit the review, sorry.

File: main/src/cgeo/geocaching/models/Waypoint.java [hunk 1]
```diff
@@ -479,7 +480,7 @@ private static WaypointType parseWaypointType(final String input, final String l
             }
         }
         for (final WaypointType wpType : WaypointType.ALL_TYPES) {
-            if (lowerInput.contains(wpType.getL10n().toLowerCase(Locale.getDefault()))) {
+            if (lowerInput.contains(wpType.getNameForNewWaypoint().toLowerCase(Locale.getDefault())) || lowerInput.contains(wpType.getL10n().toLowerCase(Locale.getDefault()))) {
```
[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Waypoint.java:483]: Hm, shouldn't we use different "newWaypointNames" for the two waypoint types then (besides doing two separate loops, if we stay with "stage" as one of them)?

I already had a collision at another place (filling the suggestion view adapter in EditWaypointActivity), but that was easy to mitigate.

[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Waypoint.java:483]: e. g.: "STAGE" (for physical stages) and "VIRTUAL"?

**Comment by moving-bits** [comment]
> Seems I forgot to submit the review, sorry.

No problem, and thanks for reviewing.

File: main/src/cgeo/geocaching/models/Waypoint.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/models/Waypoint.java:483]: I thought about this topic a lot when I implemented the whole PN parsing issue many month ago.

Imho:
I think it is close to impossible to control this in all translations. Also, interactions between subparts of the different translations would have to be considered. I don't think the "parse PN" use case is important enough to justify such strict manners.

But it should be understandable by the user what is happening, and she should have at least one possibility to specify each waypoint type uniquely in the PN (hence the introduction of the short waypoint type in parenthesis, this will always work uniquely).

In summary, I think we should make a best effort to make this work as expected by the user but not be too strict about it.

File: main/src/cgeo/geocaching/models/Geocache.java [hunk 2]
```diff
@@ -1645,7 +1645,7 @@ private Waypoint findWaypoint(final Waypoint searchWp) {
             if (!StringUtils.isBlank(searchWpName)) {
                 final String wpName = waypoint.getName();
                 final WaypointType wpType = waypoint.getWaypointType();
-                if (searchWpName.equals(wpName) && searchWp.getWaypointType().getL10n().equals(wpType.getL10n())) {
+                if (searchWpName.equals(wpName) && (searchWp.getWaypointType().getNameForNewWaypoint().equals(wpType.getNameForNewWaypoint()) || searchWp.getWaypointType().getL10n().equals(wpType.getL10n()))) {
```
[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Geocache.java:1648]: IMHO that should not be a problem here, as a full string comparison ("equals") is done here, whereas above it's a "contains". (But, actually, no need to compare both the old name and the new name, as it's about waypoint types, not names. => Will update that in my PR.)

But I wonder why we do not compare simply the type id instead of type names with each other? something like `searchWp.getWp.getWaypointType().id.equals(wpType.id)` - any reason not to?

File: main/src/cgeo/geocaching/models/Waypoint.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Waypoint.java:483]: @eddiemuc 
Do I understand that correctly that you propose to only split up the loop into two, checking the old (longer) names first, and leave the naming as it is in the PR?

[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/models/Waypoint.java:483]: > @eddiemuc 
> Do I understand that correctly that you propose to only split up the loop into two, checking the old (longer) names first, and leave the naming as it is in the PR?

Yes, that would be my suggestion

[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Waypoint.java:483]: Ok, thanks, have updated the code accordingly

File: main/src/cgeo/geocaching/models/Geocache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/models/Geocache.java]: @moving-bits you are right, didn't see that. I also don't understand why names are compares instead of ids. I think this would be better actually. @murggel any reason for this?

[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Geocache.java]: In the PR I have reverted to the original code for this method for now. If no one objects, I can switch comparison to id, and do some other optimization (eg: why do a `searchWp.getName()` within the loop etc.) - not that important, as it's a small loop, just stumbled upon while checking for waypoint type name requests.

[the remark this answers is not in the document]
Comment (murggel) [reply main/src/cgeo/geocaching/models/Geocache.java]: To be honest, I don't know anymore, why I used the name there and not the id... probably to avoid duplicate names of waypoint, if two types have different ids, but same name?

[the remark this answers is not in the document]
Comment (moving-bits) [reply main/src/cgeo/geocaching/models/Geocache.java]: Ok, thanks for replying. This is the waypoint _type_ id, which should not change, not the waypoint id. I'll create a PR.

[the remark this answers is not in the document]
Comment (murggel) [reply main/src/cgeo/geocaching/models/Geocache.java]: Hm, I think there was a problem with waypoint-type OWN an WAYPOINT. To avoid a duplicate here, if a waypoint with same name exists, I checked the names. But probably this check should be done differently.. Different waypoint-type-id would be not sufficient ...
